qemu-vm: support TPM usecases

RaitoBezarius opened the pull request:
###### Description of changes

Thanks to @ElvishJerricco, this PR enables proper TPM support inside the QEMU infrastructure.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
nixos/qemu-vm: add `virtualisation.tpm` for running TPM in QEMU infrastructure

RaitoBezarius committed:
nixos/tests: adopt newest TPM support in test infra

RaitoBezarius commented:
cc @baloo (while I have you) :).

baloo commented on nixos/lib/test-driver/test_driver/__init__.py:
```diff
@@ -77,6 +78,13 @@ def main() -> None:
         help="vlans to span by the driver",
     )
+    arg_parser.add_argument(
+        "--tpms",
+        metavar="TPMs",
+        action=EnvDefault,
+        envvar="tpms",
+        help="tpms blob to initialize by the driver (in JSON)",
```
```suggestion
        "--tpm-instances",
        metavar="TPM_INSTANCES",
        action=EnvDefault,
        envvar="TPM_INSTANCES",
        help="tpm instances to initialize by the driver (JSON encoded)",
```

baloo commented on nixos/lib/test-driver/test_driver/__init__.py:
```diff
@@ -101,6 +109,7 @@ def main() -> None:
     with Driver(
         args.start_scripts,
         args.vlans,
+        json.loads(args.tpms),
```
```suggestion
        json.loads(args.tpm_instances),
```

baloo commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -35,31 +46,44 @@ class Driver:
     tests: str
     vlans: List[VLan]
     machines: List[Machine]
+    tpms: Dict[str, Tpm]
```
```suggestion
    tpm_instances: Dict[str, Tpm]
```

baloo commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -35,31 +46,44 @@ class Driver:
         self,
         start_scripts: List[str],
         vlans: List[int],
+        tpms: List[Dict[str, str]],
```
```suggestion
        tpm_instances: List[Dict[str, str]],
```

baloo commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -35,31 +46,44 @@ class Driver:
+                tpm_data["machine_name"]: Tpm(
+                    tpm_data["swtpm_binary_path"], tpm_data["socket_path"]
+                )
+                for tpm_data in tpms
```
```suggestion
                for tpm_data in tpm_instances
```

baloo commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -35,31 +46,44 @@ class Driver:
             self.vlans = [VLan(nr, tmp_dir) for nr in vlans]
 
+        with rootlog.nested("start all TPMs"):
+            self.tpms = {
```
```suggestion
            self.tpm_instances = {
```

baloo commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -35,31 +46,44 @@ class Driver:
+                )
+                for tpm_data in tpms
+            }
+            for tpm in self.tpms.values():
```
```suggestion
            for tpm in self.tpm_instances.values():
```

baloo commented on nixos/lib/test-driver/test_driver/machine.py:
```diff
@@ -318,6 +319,8 @@ class Machine:
     shell: Optional[socket.socket]
     serial_thread: Optional[threading.Thread]
 
+    tpm: Tpm | None
```
```suggestion
    tpm: Optional[Tpm]
```
Not a python typing expert here, but that's what `Optional` is about right?

nikstur replied:
The Union `|` Operator is newer than `Optional`. It was introduced in Python 3.10 ([PEP 604](https://peps.python.org/pep-0604/)) and I'd wager, nowadays, `Tpm | None` is even more idiomatic than the older `Optional[Tpm]`.

baloo commented on nixos/lib/test-driver/test_driver/tpm.py:
```diff
@@ -0,0 +1,54 @@
+                "--tpm2",
+            ]
+        )
+
```
I'd poll for `self.tpm_socket_path` to appear for one or two seconds. Just to make sure we don't hand that over to qemu before it's ready.

ElvishJerricco replied:
`swtpm` does have a `--daemon` flag. I wonder if it does that correctly, where the parent process only exits successfully once it knows the child is running and ready for use. If so, rather than polling, and rather than the little `0.2` second timeout below, we could just use `--daemon` and wait for a successful exit code from the parent.

baloo commented on nixos/lib/testing/driver.nix:
```diff
@@ -14,6 +14,14 @@ let
 
   vlans = map (m: m.virtualisation.vlans) (lib.attrValues config.nodes);
   vms = map (m: m.system.build.vm) (lib.attrValues config.nodes);
+  tpms = map (n:
```
```suggestion
  tpm_instances = map (n:
```

baloo commented on nixos/lib/testing/driver.nix:
```diff
@@ -98,6 +114,7 @@ let
           --set startScripts "''${vmStartScripts[*]}" \
           --set testScript "$out/test-script" \
           --set vlans '${toString vlans}' \
+          --set tpms '${builtins.toJSON tpms}' \
```
```suggestion
          --set tpm_instances '${builtins.toJSON tpm_instances}' \
```

ElvishJerricco commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -796,6 +796,42 @@ in
       };
     };
 
+    virtualisation.tpm = {
```
Seems odd to have this in `qemu-vm.nix` when it requires the test driver. Either `test-instrumentation.nix` or an implementation that isn't dependent on the driver seems preferable (with the latter seeming fairly impractical).

RaitoBezarius replied:
Good point.

RaitoBezarius replied:
A driver-less implementation could be provided by modifying the QEMU run script and adding a parallel program to run `swtpm` actually.

tfc commented on nixos/lib/test-driver/test_driver/__init__.py:
```diff
@@ -120,7 +129,7 @@ def generate_driver_symbols() -> None:
     scripts.
     """
-    d = Driver([], [], "", Path())
+    d = Driver([], [], [], "", Path())
```
hm we might want to switch to name the parameters to make this less messy. Can you please do it at this occasion?

tfc commented:
Hi Ryan,

I just skimmed through the code. It feels like the TPM support that you just built would also be very very useful for everyone who wants to use it with the sole `run-qemu-vm` script without the test driver.
What do you think about also supporting those by making the whole TPM thing and daemon launching and socket preparation a part of the qemu VM bash script?
Just configuring this in the VM's NixOS configuration and then having it provided by the sole bash script would be very elegant and powerful.

This way we would have support for mentioned single-vm-without-test-driver scenarios, and we could drop all of the python code changes (Users just need to "find" the socket either by convention or adding additional parameters to the machine class so they reach the vm script, we can still streamline that when we have actual test that really use it).

What do you think?

RaitoBezarius commented:
> Hi Ryan,
> 
> I just skimmed through the code. It feels like the TPM support that you just built would also be very very useful for everyone who wants to use it with the sole `run-qemu-vm` script without the test driver. What do you think about also supporting those by making the whole TPM thing and daemon launching and socket preparation a part of the qemu VM bash script? Just configuring this in the VM's NixOS configuration and then having it provided by the sole bash script would be very elegant and powerful.

Very good point. I will try to go for this change. :)
Thank you!

> 
> This way we would have support for mentioned single-vm-without-test-driver scenarios, and we could drop all of the python code changes (Users just need to "find" the socket either by convention or adding additional parameters to the machine class so they reach the vm script, we can still streamline that when we have actual test that really use it).
> 
> What do you think?

Ideally, I'd love to have it still "very automatic" like you can see it in the systemd credentials test for QEMU test infra, but I am okay with going incrementally on that!

nikstur commented:
Support for TPMs in a "normal" non-test VM would be awesome.

SuperSandro2000 commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -796,6 +796,42 @@ in
+
+      socketPath = mkOption {
+        type = types.str;
+        default = "/tmp/swtpm-sock";
```
```suggestion
        default = "/run/swtpm-sock";
```

SuperSandro2000 commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -796,6 +796,42 @@ in
+          "ppc64-linux" = "tpm-spapr";
+          "armv7-linux" = "tpm-tis-device";
+          "aarch64-linux" = "tpm-tis-device";
+        }.${pkgs.hostPlatform.system});
```
This should get an or to catch incompatible arches early

ElvishJerricco commented:
I've made it so that the implementation is entirely in `qemu-vm.nix`, and I've updated a couple more tests to use it.

Question at the moment: In my testing, `qemu` does reliably shut down the TPM itself. So I don't have any code to ensure the `swtpm` process is exited. Is this acceptable? Or do we need to do something about it?

RaitoBezarius commented:
> I've made it so that the implementation is entirely in `qemu-vm.nix`, and I've updated a couple more tests to use it.
> 
> Question at the moment: In my testing, `qemu` does reliably shut down the TPM itself. So I don't have any code to ensure the `swtpm` process is exited. Is this acceptable? Or do we need to do something about it?

I think it's okay for now, I just hope swtpm cannot enter into an uninterruptible sleep mode or something.

ElvishJerricco commented:
@RaitoBezarius Actually, I realized a couple cases where we can't rely on that. 1) If our bash script fails before `exec qemu`, but this is fixable with `trap ... EXIT`. But if qemu fails to start after being exec'd, e.g. because it rejects the cmdline args, it probably won't shutdown swtpm.

baloo committed:
qemu-vm: stop the swtpm once qemu stops

The idea is to run an async process waiting for swtpm
and we have to ensure that `FD_CLOEXEC` is cleared on this process'
stdin file descriptor, we use `fdflags` for this, a loadable builtin in
Bash ≥ 5.

The async process when exited will terminate `swtpm`, we bind the
termination of the async process to the termination of QEMU by virtue of
having `qemu` exec in that Bash script.

Signed-off-by: Arthur Gautier <baloo@superbaloo.net>
Co-authored-by: Raito Bezarius <masterancpp@gmail.com>

baloo commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -198,6 +198,30 @@ let
+        # which will stops it.
+        coproc waitingswtpm {
+          read || :
+          echo "" | ${lib.getExe hostPkgs.socat} STDIO UNIX-CONNECT:"$NIX_SWTPM_DIR"/socket
```
```suggestion
         # Bash will set the `FD_CLOEXEC` flag on the pipe from parent to coproc
         # this will trigger the coproc to read EOF when qemu is `exec` below.
         # To work around that, we will wait a second before triggering swtpm to exit, assuming qemu will have
         # already connected to swtpm by then.
          sleep 1
          echo "" | ${lib.getExe hostPkgs.socat} STDIO UNIX-CONNECT:"$NIX_SWTPM_DIR"/socket
```

Got to re-read an strace of bash behavior in trying to reply to @ElvishJerricco (https://github.com/NixOS/nixpkgs/pull/246507#issuecomment-1773859259), discovered I missed the `fcntl(63, F_SETFD, FD_CLOEXEC) = 0` the parent runs.

<details>
<summary>
strace output // analysis
</summary>

``` bash
# foo.sh
coproc waitparent {
   read || :
   echo "bar"
}

sleep 1

echo "foo"

exec sleep 2
```

strace output:
```
$ strace -t -f -e clone,dup2,pipe2,write,read,close,fcntl bash foo.sh
[...]
<<< Parent creates a pipe and rebinds parts to FD 63 & 62 // coproc -> parent pipe >>>
21:45:37 pipe2([3, 4], 0)               = 0
21:45:37 dup2(3, 63)                    = 63
21:45:37 close(3)                       = 0
21:45:37 dup2(4, 62)                    = 62
21:45:37 close(4)                       = 0
<<< Parent creates a pipe and rebinds parts to FD 61 & 6 // parent -> coproc pipe >>>
21:45:37 pipe2([3, 4], 0)               = 0
21:45:37 dup2(3, 61)                    = 61
21:45:37 close(3)                       = 0
21:45:37 dup2(4, 60)                    = 60
21:45:37 close(4)                       = 0
<<< Parent spawns the coproc >>>
21:45:37 clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f72dd92fa10) = 1849390
strace: Process 1849390 attached
<<< Parent closes the write part of the coproc -> parent pipe >>>
[pid 1849389] 21:45:37 close(62)        = 0
<<< Parent closes the read part of the parent -> coproc pipe >>>
[pid 1849389] 21:45:37 close(61)        = 0
<<< Parent sets CLOEXEC on its FDs for the pipe >>>
[pid 1849389] 21:45:37 fcntl(63, F_SETFD, FD_CLOEXEC) = 0
[pid 1849389] 21:45:37 fcntl(60, F_SETFD, FD_CLOEXEC) = 0
<<< Child closes the read part of the coproc -> parent pipe >>>
[pid 1849390] 21:45:37 close(63)        = 0
<<< Child closes the write part of the parent -> coproc pipe >>>
[pid 1849390] 21:45:37 close(60)        = 0
<<< Child rebinds the pipe to stdin (FD 0) and stdout (FD 1)
[pid 1849390] 21:45:37 dup2(61, 0)      = 0
[pid 1849390] 21:45:37 close(61)        = 0
[pid 1849390] 21:45:37 dup2(62, 1)      = 1
[pid 1849390] 21:45:37 close(62)        = 0
<<< Child starts reading from the pipe >>>
[pid 1849390] 21:45:37 read(0,  <unfinished ...>
<<< Parent launches the `sleep 1` command >>>
[pid 1849389] 21:45:37 clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f72dd92fa10) = 1849391
strace: Process 1849391 attached
[pid 1849391] 21:45:38 +++ exited with 0 +++
[pid 1849389] 21:45:38 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1849391, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
<<< Sleep exits after 1 sec >>>
[pid 1849389] 21:45:38 write(1, "foo\n", 4) = 4

<<< Parent will now exec `sleep 2` >>>
<<< Child read now reads EOF >>>
[pid 1849390] 21:45:38 <... read resumed>"", 1) = 0
[pid 1849390] 21:45:38 write(1, "bar\n", 4) = -1 EPIPE (Broken pipe)
[pid 1849390] 21:45:38 --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=1849390, si_uid=1000} ---
[pid 1849389] 21:45:38 close(3)         = 0
[pid 1849390] 21:45:38 +++ killed by SIGPIPE +++
21:45:38 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=1849390, si_uid=1000, si_status=SIGPIPE, si_utime=0, si_stime=0} ---
21:45:40 +++ exited with 0 +++
```

</details>

RaitoBezarius replied:
Can we avoid the sleep?

RaitoBezarius replied:
(If I'm not wrong, on hydra.nixos.org, this has the potential to exercise spurious failures under heavy load…)

RaitoBezarius replied:
Found a way, using `fdflags`.

RaitoBezarius commented:
I tested locally the 3 TPM2 tests, they are still passing and any non-TPM2 tests should still be passing with this, @ElvishJerricco do we have any blocker or concern over this PR?

flokli commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -204,8 +204,31 @@ let
+        enable -f ${hostPkgs.bash}/lib/bash/fdflags fdflags
+        # leave a dangling subprocess because the swtpm ctrl socket has
+        # "terminate" when the last connection disconnects, it stops swtpm.
+        # When qemu stop, or if the main shell process ends, the coproc will
```
When qemu /stops/

ElvishJerricco reviewed (approved):
@RaitoBezarius The only concern I had was with stopping swtpm, which is now addressed. So this looks good to me.
